lib/python/flame/registry/local.py

```diff
@@ -0,0 +1,124 @@
+from ..config import Config
+import logging
+import os
+import csv
+import pickle
+from collections import defaultdict
+from flame.common.util import (get_ml_framework_in_use, MLFramework)
+
+logger = logging.getLogger(__name__)
+
+class LocalRegistryClient(AbstractRegistryClient):
+    """Local registry client."""
+
+    _instance = None
+
+    def __new__(cls):
+        """Create a singleton instance."""
+        if cls._instance is None:
+            logger.info("Create a local registry client instance")
+            cls._instance = super().__new__(cls)
+        return cls._instance
+    
+    def __call__(self, uri: str, config: Config) -> None:
+        """Initialize the instance."""
+        self.job_id = config.job.job_id
+        self.task_id = config.task_id
+
+    def setup_run(self, name: str) -> None:
+        """Set up a run."""
+        # set up directories
+        home_dir = os.path.expanduser("~")
+        log_dir = os.path.join(home_dir, "flame-log")
+        self.registry_path = os.path.join(log_dir, self.job_id, self.task_id)
+        
+        os.system(f'rm -rf {self.registry_path}')
+        os.makedirs(self.registry_path)
+        os.mkdir(os.path.join(self.registry_path, "metrics"))
+        os.mkdir(os.path.join(self.registry_path, "model"))
+        os.mkdir(os.path.join(self.registry_path, "params"))
+        #os.mkdir(os.path.join(self.registry_path, "artifact"))
+        
+        # version tracking
+        self.param_version = 1
+        self.model_versions = defaultdict(lambda: 1)
+
+    def save_metrics(self, epoch: int, metrics: Optional[dict[str,
+                                                              float]]) -> None:
+        """Save metrics in a model registry."""
+        for metric in metrics:
+            filename = os.path.join(self.registry_path, "metrics", metric)
+            if os.path.exists(filename):
+                with open(filename, 'a') as file:
+                    csv_writer = csv.writer(file)
+                    csv_writer.writerow([epoch, metrics[metric]])
+            else:
+                with open(filename, 'w') as file:
+                    csv_writer = csv.writer(file)
+                    csv_writer.writerow(["round", metric])
+                    csv_writer.writerow([epoch, metrics[metric]])
+
+    def save_params(self, hyperparameters: Optional[Hyperparameters]) -> None:
+        """Save hyperparameters in a model registry."""
+        with open(os.path.join(self.registry_path, "params", str(self.param_version)), 'wb') as file:
+            pickle.dump(hyperparameters.dict(), file)
+        self.param_version += 1
+
+    def cleanup(self) -> None:
+        """Clean up resources."""
+        pass
+
+    def save_model(self, name: str, model: Any) -> None:
+        """Save a model in a model registry."""
+        model_folder = os.path.join(self.registry_path, "model", name)
+        if not os.path.exists(model_folder):
+            os.mkdir(model_folder)
```

Comment:
```suggestion
            os.makedirs(model_folder, exist_ok=True)
```

You may consider merging 95 and 96 into one line like this